Add EigenDA V2 CI Infrastructure

shyam-patel-kira opened the pull request:
### Add E2E Tests and Disperser Mode Support for EigenDA V2

- Implements dual-mode testing for EigenDA V2 integration

shyam-patel-kira committed:
Add EigenDA V2 CI infrastructure

shyam-patel-kira committed:
Add ReferenceDA system tests and update V2 CI workflow

shyam-patel-kira committed:
Improve ReferenceDA tests with certificate verification and comprehensive documentation

shyam-patel-kira committed:
Enable EigenDA V2 with comprehensive system and e2e tests

shyam-patel-kira committed:
Fix double http:// prefix bug in ReferenceDA test URLs

shyam-patel-kira committed:
Fix lint errors: add missing JWTSecret field to ServerConfig

shyam-patel-kira committed:
Fix errcheck lint errors: properly handle server.Shutdown error returns

shyam-patel-kira committed:
Handle V2 certificates: trigger failover when V2 detected

shyam-patel-kira committed:
Restore EigenDA V2 + ReferenceDA failover test configuration

ethenotethan commented on scripts/start-eigenda-proxy.sh:
```diff
@@ -1,34 +1,83 @@
+    DISPERSAL_BACKEND="V2"
+    # Enable admin API for runtime backend switching
+    ENABLE_ADMIN_API="true"
+    ;;
```
what exactly are you testing? I'd imagine we'd need to configure these server specific [env vars](https://github.com/Layr-Labs/eigenda/blob/master/api/proxy/docs/help_out.txt#L16-L35) as well as pass in `arb` as a supported option via the `APIS_ENABLED` user [flag](https://github.com/Layr-Labs/eigenda/blob/master/api/proxy/docs/help_out.txt#L239-L242)

shyam-patel-kira replied:
Yes, that is correct, the latest changes I pushed has updated this script

shyam-patel-kira committed:
Deploy ReferenceDAProofValidator contract in tests and add ALT DA support

This commit fixes ReferenceDA certificate validation failures by:

1. **Contract Support for ALT DA Certificates (0x01)**:
   - Added DAC_CERTIFICATE_MESSAGE_HEADER_FLAG (0x01) to SequencerInbox.sol
   - Updated isValidCallDataFlag() to accept 0x01 header byte
   - Added flag to ISequencerInbox.sol interface
   - Added DACertificateMessageHeaderFlag to KnownHeaderBits in daprovider/util.go

2. **ReferenceDAProofValidator Contract Deployment**:
   - Modified setupReferenceDAServer() to deploy ReferenceDAProofValidator contract with trusted signer
   - Updated setupReferenceDAServerForFallback() with same deployment logic
   - Added l1info parameter to both functions to get transaction options for deployment
   - Added import for solgen/go/localgen to access contract deployment bindings

3. **Test Improvements**:
   - Both TestReferenceDAIntegration and TestEigenDAV2WithReferenceDAFallback now pass
   - Certificate signer is properly registered as a trusted signer in deployed validator contract
   - Contract address is used for on-chain validation during certificate verification

Previous errors resolved:
- ❌ "certificate validation failed: no contract code at given address" → ✅ Fixed
- ❌ "InvalidHeaderFlag(0x01)" from SequencerInbox contract → ✅ Fixed

Tests passing locally:
- ✅ TestReferenceDAIntegration (22.12s)
- ✅ TestEigenDAV2WithReferenceDAFallback (22.07s)

shyam-patel-kira committed:
Update contracts to eigenda-v2.1.3-altda-support with ALT DA support

shyam-patel-kira committed:
Update contracts to v3.9.0-altda-support with ALT DA support

shyam-patel-kira committed:
Add e2e tests and disperser mode support for EigenDA V2 with arb API

shyam-patel-kira committed:
Remove placeholder tests and consolidate TODOs

ethenotethan commented on contracts:
why is the submodule changed?

shyam-patel-kira replied:
it was updated to add `DAC_CERTIFICATE_MESSAGE_HEADER_FLAG` (0x01) support in SequencerInbox.sol. This allows the contract to accept ReferenceDA certificates which use the ALT-DA spec. Without this, ReferenceDA batches are rejected as invalid.

ethenotethan replied:
still a bit fuzzy why contract submodules need to be updated here at all given the upstream repo has golang tests for their CustomDA feature which asserts to correctness of batch posting / derivation logics

shyam-patel-kira replied:
Yes, upstream has CustomDA support and the golang daprovider/ code already supports 0x01 (we inherit that). However, the L1 contracts must also accept 0x01 headers for batch posting to work. Without this update, batch posting would fail with InvalidHeaderFlag(0x01).

The update adds one line to SequencerInbox.sol to validate the 0x01 header byte - same as upstream's CustomDA support, just using our fork's naming (`DAC_CERTIFICATE_MESSAGE_HEADER_FLAG` vs upstream's `CUSTOM_DA_MESSAGE_HEADER_FLAG`).

ethenotethan commented on scripts/start-eigenda-proxy.sh:
```diff
@@ -1,34 +1,139 @@
+# Version parameter: v1 or v2 (default: v1)
+VERSION="${1:-v1}"
+# Mode parameter: memstore or disperser (default: memstore)
+MODE="${2:-memstore}"
```
why are we supporting version switching? I thought the idea was to nuke all V1 codepaths entirely

shyam-patel-kira replied:
yea, I kept it in case we wanted to do an incremental rollout, can nuke it completely, don't have a strong opinion on this one

ethenotethan replied:
vs overloading the existing `start-eigenda-proxy.sh` why not add a separate one `start-eigenda-proxy-customda.sh` to easier support followup deletion of V1 related logics?

shyam-patel-kira replied:
that's a good idea, create a different setup script for v2 testing: [a37f34d](https://github.com/Layr-Labs/nitro/pull/144/commits/a37f34d35746b76b907a26da4af1a9124777ba41)

ethenotethan commented on eigenda/proxy.go:
```diff
@@ -28,10 +28,25 @@ func (c *EigenDAProxyClient) Put(ctx context.Context, data []byte) (*disperser.B
 		return nil, fmt.Errorf("failed to set data: %w", err)
 	}
 
+	if len(cert) == 0 {
+		return nil, fmt.Errorf("received empty certificate from proxy")
+	}
+
+	// Check version byte to determine certificate format
+	version := cert[0]
+
+	// V2 certificate (version 0x02): Not supported through V1 code path
+	// V2 uses ALT-DA spec and should be accessed through DAProvider interface, not EigenDA.Enable
+	// Returning ErrServiceUnavailable will trigger failover to DAProvider if configured
+	if version == 0x02 {
+		return nil, standard_client.ErrServiceUnavailable
+	}
```
why is this check being done?

shyam-patel-kira replied:
This detects V2 certificates in the V1 code path and returns ErrServiceUnavailable to trigger failover to ReferenceDA. However, if we're removing V1 entirely, this check may no longer be needed.

ethenotethan replied:
1. this implementation deviates from the initial spec [proposal](https://www.notion.so/eigen-labs/V1-V2-Nitro-Branch-Management-Strategy-29a13c11c3e08023bb8ef19aa2c7db24?source=copy_link) where we'd have a separate branch that doesn't contain any V1 code changes or we'd nuke v1 code paths. If we're doing the approach of cannibalizing the existing `eigenda` branch then we'll need a followup PR that nukes all V1 references.
2. V2 uses `CustomDA` which is provided for us by OCL. These EigenDA specific write/read system paths would never be touched for EigenDA V2 x CustomDA and leverage the existing `CustomDA writer/reader` provided to us in the software.

ethenotethan commented on system_tests/eigenda_v2_e2e_test.go:
```diff
@@ -0,0 +1,137 @@
+	proxyV2E2EURL = "http://127.0.0.1:4242"
+)
+
+// TestEigenDAV2E2EConnectivity verifies the proxy is properly configured for real disperser
```
why do we need this test right now? this PR feels like its scope creep limits. I'd prefer if we more incremental with this rollout where we first introduce just `memstore` vs having context switching for different EigenDA backends

ethenotethan commented on system_tests/eigenda_v2_e2e_test.go:
```diff
@@ -0,0 +1,137 @@
+	// Setup second node for sync testing
+	l1NodeConfigB := arbnode.ConfigDefaultL1NonSequencerTest()
+	l1NodeConfigB.BlockValidator.Enable = false
+	l1NodeConfigB.EigenDA.Enable = true
```
why are we touching this configuration?

shyam-patel-kira replied:
We're configuring both because this PR tests EigenDA V2 (primary) with ReferenceDA as fallback. The `EigenDA.Enable` config points to the V2 proxy using ALT-DA spec, while DAProvider provides the fallback mechanism. However if we want to do an incremental rollout I can remove it, not sure if this scope is too large.

ethenotethan replied:
> The EigenDA.Enable config points to the V2 proxy using ALT-DA spec

right V2 proxy uses the *ALT-DA spec* which assumes the CustomDA server feature of proxy which lives independent of the REST API that this `EigenDA` config expects communication with. All nitro client related communication paths for CustomDA write/read are supported in the software: 
https://github.com/Layr-Labs/nitro/blob/eigenda/daprovider/daclient/daclient.go

shyam-patel-kira replied:
[a37f34d](https://github.com/Layr-Labs/nitro/pull/144/commits/a37f34d35746b76b907a26da4af1a9124777ba41)

shyam-patel-kira committed:
Remove V1 codepaths and simplify to V2-only through DAProvider interface

ethenotethan commented on system_tests/referenceda_test.go:
```diff
@@ -0,0 +1,304 @@
+		data_streaming.PayloadCommitmentVerifier(),
+	)
+	Require(t, err)
+
```
what server are you spinning up here?

shyam-patel-kira replied:
it doesn't exist anymore but this was for referenceDA server we're not touching it anymore

shyam-patel-kira committed:
Address review: remove V1 modifications, separate V2 script, drop ReferenceDA tests

shyam-patel-kira committed:
Handle missing GetSupportedHeaderBytes RPC method gracefully for simple DA proxies

shyam-patel-kira committed:
Fix V2 proxy startup to use locally-built binary with Arbitrum RPC support

Signed-off-by: Shyam Patel <shyampkira@gmail.com>

shyam-patel-kira committed:
Use Docker-based proxy build for V2 with Arbitrum RPC support

Signed-off-by: Shyam Patel <shyampkira@gmail.com>

shyam-patel-kira committed:
Exclude V2 proxy Dockerfile from main Docker build context
